appengine/reconciletags/version_check_test.py

```diff
@@ -7,46 +7,129 @@
+    "debian": "apt-get update; apt-get -y install lsb-release; lsb_release -a",
+    "ubuntu": "apt-get update; apt-get -y install lsb-release; lsb_release -a",
+    "ruby": "ruby -v",
+    "python": "python3 --version",
+    "php": "php -v",
+    "nodejs": "node --version",
+    "java": "java -version",
+    "go1-builder": "echo $GO_VERSION"
 }
 
 
 class VersionCheckTest(unittest.TestCase):
+    def filter_node(s, current, version=None):
+        if current:
+            return s.lstrip('v').rstrip()
+        else:
+            return re.findall(r'v({}.\d+)'.format(version), s)[0]
+
+    def filter_python(s, current, version=None):
+        if current:
+            return s.split()[1]
+        else:
+            return re.findall(r'{}.\d+'.format(version), s)[-1]
+
+    def filter_ruby(s, current, version=None):
+        if current:
+            return s.split()[1][:-4]
+        else:
+            return re.findall(r'Ruby ({}.\d+)'.format(version), s)[1]
+
+    def filter_php(s, current, version=None):
+        if current:
+            return s.split()[1]
+        else:
+            return re.findall(r'PHP ({}.\d+)'.format(version), s)[0]
+
+    def filter_debian(s, current, version=None):
+        if current:
+            return re.findall(r'Description:[\s|\S]+(\d+.\d+)', s)[0]
+        else:
+            return re.findall(r'Debian ({}.\d+)'.format(version), s)[1]
 
-    def _get_latest_version(self, runtime, version):
-        cmd = (runtime_to_version_check.get(runtime)
-               .format(version.replace('.', '\\.')))
+    def filter_ubuntu(s, current, version=None):
+        if current:
+            return re.findall(r'Description:[\s|\S]+(\d\d+.\d+.\d+)', s)[0]
+        else:
+            return re.findall(r'Ubuntu ({}.\d+)'.format(version), s)[1]
+
+    def filter_aspnetcore(s, current, version=None):
+        if current:
+            return re.findall(r'Version: (\d+.\d+.\d+)', s)[0]
+        else:
+            v = re.findall(r'v({}.\d+)'.format(version), s)
+            v.sort(key=LooseVersion)
+            return v[-1]
+
+    def filter_java(s, current, version=None):
+        if current:
+            return re.findall(r'OpenJDK Runtime Environment '
+                              '\(build \S+_\d+-(\S+)-\S{3}', s)[0]
+        else:
+            return re.findall(r'Selected version \'(\S+)\'', s)[0]
+
+    def filter_go(s, current, version=None):
+        if current:
+            return s.rstrip()
+        else:
+            return re.findall(r'go({}.\d+)'.format(version), s)[0]
+
+    runtime_to_filter = {
+        "debian": filter_debian,
+        "ubuntu": filter_ubuntu,
+        "php": filter_php,
+        "nodejs": filter_node,
+        "python": filter_python,
+        "java": filter_java,
+        "ruby": filter_ruby,
+        "aspnetcore": filter_aspnetcore,
+        "go1-builder": filter_go
+    }
+
+    def _get_latest_version(self, runtime, version, image):
+        if runtime == 'java':
```

Comment:
Why does java need a special case here? Can you add a comment?

Reply:
It's the only language that needs image and version passed into the command, but turns out we can vacuously pass in those variables and just not use them for everyone else. The special case is gone.

Reply:
Nice, much cleaner now.